src/build/_util.py

```diff
@@ -53,6 +53,12 @@ def check_dependency(
         # dependency is not installed in the environment.
         yield (*ancestral_req_strings, normalised_req_string)
     else:
+        if dist.version is None:
+            # Malformed dist-info or some other form of metadata corruption
+            # req.specifier.contains will raise TypeError, so let's do the same
+            # with a more helpful error message
+            msg = f'Package {req.name} has malformed metadata and no version information could be found'
+            raise TypeError(msg)
```

Comment:
I wonder if we shouldn't make this a warning instead and return this package as an unmet dependency. That should give more information to the user. As it stands, they would only see an error message about a specific bad dependency, instead of getting a list of all bad ones with the extra context of why this one was considered bad.

Reply:
Seems fine to me, sounds reasonable to want all issues at the same time instead of potentially having to address one-by-one